### PR TITLE
[senechome] Avoid null pointer errors

### PR DESCRIPTION
**Pull request by thopiekar**
Looks like the new generation of SenecIES power storage systems, which are using LiPo technology, provide additional data in the JSON array.
This data is not available in the old generation, which lead to missing data, when asking for a key and in further null pointer exceptions when going deeper in the data structure.

Therefore making a trivial check to detect the new generation of devices and skip the troubling section entirely.
CON: The section of items would not be updated. A to-do is of course to split the old and new generation into separate models of devices.

Fixes https://github.com/openhab/openhab-addons/issues/11606

Also-By: Erwin Guib <eguib@web.de>
Signed-off-by: Thomas Karl Pietrowski <thopiekar@gmail.com>
Signed-off-by: Korbinian Probst <kp.droid.dev@gmail.com>

### CODE REVIEW

**Comment by thopiekar** [comment]
*NOTE*: Needs to be tested. Theoretically, it should work but coded all with my best gut feeling!

I'm currently on a milestone release and getting this error:
```
2022-03-23 21:16:13.646 [WARN ] [internal.service.FeaturesServiceImpl] - Can't load features repository mvn:org.apache.karaf.features/framework/4.3.2/xml/features
java.lang.RuntimeException: Error resolving artifact org.apache.karaf.features:framework:xml:features:4.3.2: [Could not find artifact org.apache.karaf.features:framework:xml:features:4.3.2 in openhab (https://openhab.jfrog.io/openhab/libs-milestone/)] : mvn:org.apache.karaf.features/framework/4.3.2/xml/features
        at org.apache.karaf.features.internal.service.RepositoryImpl.load(RepositoryImpl.java:121) ~[bundleFile:?]
        at org.apache.karaf.features.internal.service.RepositoryImpl.<init>(RepositoryImpl.java:51) ~[bundleFile:?]
```

No idea how to solve this and test my own work :disappointed:

File: bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java:216]: `response.battery` could also be null. Furthermore, some channels requiring information not coming from `response.battery.chargedEnergy` are now skipped. Perhaps this block could be extracted into a private method with `response.battery` provided as parameter, and channels could be updated according to information available? This would avoid assumptions and also reduce nesting.

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks - just a single comment with some advise to be sure problem is fixed and improve reliability. Perhaps it would also be worth checking if there are any null annotation warnings, didn't try to run a build myself.

File: bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java
Comment (thopiekar) [reply bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java:216]: Good point. I went for the path with the lesser effort first.
I would put some more time into this if I could test it myself.
Some strange error messages prevent me from updating my bindings in general. I've updated to 3.3M2, but I can't update my bindings from 3.1. Any idea?
-> https://community.openhab.org/t/openhab-3-3-milestone-discussion/132715/65?u=thopiekar

Comment (jlaur) [reply bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java:216]: Does it mean that your installation is broken after upgrading to 3.3M2? Unfortunately I don't know what could be the reason for this.

Comment (thopiekar) [reply bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java:216]: I think so, but the message doesn't give me a clue how to fix it :disappointed:

**Comment by thopiekar** [comment]
I managed to get it working just now. I had to wipe userdata/{cache,tmp}.
@jlaur Your suggestion doesn't work since there is always something inside response.battery.
There is a class that describes the expected dataset and when deserializing the received one, the ones that have not been sent will be omitted.
Therefore response.battery != null, but its children are.

I'm going to update the PR once I'm done.

File: bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java
Comment (jlaur) [reply bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java:216]: @thopiekar - just to be clear, intention was something like this:

```
if (response.battery != null) {
    updateBatteryStates(response.battery);
}
...

private void updateBatteryStates(SenecHomeBattery battery) {
    if (battery.chargedEnergy != null) {
        updateQtyState(CHANNEL_SENEC_CHARGED_ENERGY_PACK1, response.battery.chargedEnergy[0], 2, Units.KILOWATT_HOUR, DIVISOR_MILLI_TO_KILO);
        ....
    }
    if (battery.dischargedEnergy != null) {
        updateQtyState(CHANNEL_SENEC_DISCHARGED_ENERGY_PACK1, response.battery.dischargedEnergy[0], 2,
                        Units.KILOWATT_HOUR, DIVISOR_MILLI_TO_KILO);
        ...
    }
}
```

Remember also to check array sizes before indexing.

**Comment by jlaur** [comment]
> Your suggestion doesn't work since there is always something inside response.battery.

I understand, but that was only part of the suggestion. Please see https://github.com/openhab/openhab-addons/pull/12512#discussion_r835964327.

File: bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java
Comment (thopiekar) [reply bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java:216]: But response.battery will never get null. As I said, the "skeleton" of the data is described in json/SenecHomeResponse.java as classes.
Even if my device doesn't send any battery info in its response, there will be empty children set by the class structure.
Therefore this might be an option:
```
    updateBatteryStates(response.battery);
...

private void updateBatteryStates(SenecHomeBattery battery) {
    if (battery.chargedEnergy != null) {
        updateQtyState(CHANNEL_SENEC_CHARGED_ENERGY_PACK1, battery.chargedEnergy[0], 2, Units.KILOWATT_HOUR, DIVISOR_MILLI_TO_KILO);
        ....
    }
    if (battery.dischargedEnergy != null) {
        updateQtyState(CHANNEL_SENEC_DISCHARGED_ENERGY_PACK1, battery.dischargedEnergy[0], 2, Units.KILOWATT_HOUR, DIVISOR_MILLI_TO_KILO);
        ...
    }
}
```

But this would be only a cosmetic thing in my point of view.

**Comment by thopiekar** [comment]
@eguib Want to take a look at this PR? :eyes:

File: bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java
Comment (jlaur) [reply bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java:216]: OK. The exception from the issue mentioned is addressed by this PR, so resolving comment.

**Review by jlaur — APPROVED** [review]
Thanks for fixing this.

**Comment by jlaur** [comment]
> @eguib Want to take a look at this PR? 👀

Let me know when you are ready for having it merged. :)

File: bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java
Comment (KorbinianP) [inline bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java:225]: Hey @thopiekar , how can you ensure that `dischargedEnergy` is not null? Just because `chargedEnergy` is not null? I have not digged into this. @eguib did a nice proposal here: https://github.com/openhab/openhab-addons/issues/11606#issuecomment-1078840338 , could you take this over?

**Review by KorbinianP — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java
Comment (thopiekar) [reply bundles/org.openhab.binding.senechome/src/main/java/org/openhab/binding/senechome/internal/SenecHomeHandler.java:225]: Yes, I can pick the patch and mention him here as an author, if you and @eguib like :)

**Review by jlaur — CHANGES_REQUESTED** [review]
See https://github.com/openhab/openhab-addons/pull/12512#discussion_r837390168

**Comment by thopiekar** [comment]
@jlaur We are ready to go! Feel free to take a last look and, if everything is fine, push the merge button :smile:

**Review by jlaur — APPROVED** [review]
LGTM

**Review by KorbinianP — APPROVED** [review]
Thank you for the patch

**Comment by thopiekar** [comment]
Ok, corrected the signed-of tags. Made @eguib "also-by".
Just don't see how the commit message is too long.. The title is perfectly fine (max 50 letters) and I read about no limit on the commit description.

```
>>> len("[senechome] Avoid null pointer errors")
37
```

**Comment by jlaur** [comment]
> Ok, corrected the signed-of tags. Made @eguib "also-by".
> Just don't see how the commit message is too long.. The title is perfectly fine (max 50 letters) and I read about no limit on the commit description.

It's fine! I might have mixed up something previously. It's ready to go, thanks for the fix. I can tweak the sign-off further when squash-merging as I believe there should be only one sign-off, so I would move Korbinian to "Also-By". Please confirm. After that I'll merge.

**Comment by thopiekar** [comment]
@jlaur Is there anything left? :smile:

**Comment by thopiekar** [comment]
Well, @KorbinianP didn't take a look at the problem in-depth, but reviewed @eguib 's and my solution. Therefore I would keep the sense and keep him "signed-off-by".

**Comment by thopiekar** [comment]
.. otherwise someone might get questions, look into the commit message and contact him about why the heck something has been done as is. I mean yeah, he knows the case now, but logically I would keep the commit as it is.

**Comment by jlaur** [comment]
> Well, @KorbinianP didn't take a look at the problem in-depth, but reviewed @eguib 's and my solution. Therefore I would keep the sense and keep him "signed-off-by".

For such small commit I guess the usual approach would be simply letting the committer sign-off the commit. But since you insist, I have to consult other maintainers, since I'm not fully aware of the rules. @openhab/add-ons-maintainers - is it allowed to have more than one "Signed-off-by" in commit message? The DCO check passes, but it seems inconsistent being able to mix "Signed-off-by" and "Also-by" this way - it could even obsolete "Also-by" if you could just choose to have multiple "Signed-off by" instead?

**Comment by thopiekar** [comment]
I'm not sure about the rules in detail here, too, just following my logic and my gut feeling.
If there is anything I need to correct still let me know! :+1:

**Comment by KorbinianP** [comment]
Hello, how do I sign off if I don't commit anything?

**Comment by KorbinianP** [comment]
Cool, thank you all!

**Comment by thopiekar** [comment]
Thank you for the explanation @wborn

**Pull request merged**
